### ISSUE TITLE
stage2: Sentinel not validated for comptime slice

### ISSUE THREAD

**Issue by topolarity** [ziglang/zig#12037]
### Zig Version

0.10.0-dev.2855+aab1284e1

### Steps to Reproduce

Found this while looking into #12033 (apologies if it's already covered by existing tests)

```zig
const std = @import("std");

test {
    const string = "Hello!\x00World!";
    try std.testing.expect(@TypeOf(string) == *const [13:0]u8);

    const slice_with_sentinel: [:0]const u8 = string[0..8 :0];
    try std.testing.expect(@TypeOf(slice_with_sentinel) == [:0]const u8);

    std.debug.assert(slice_with_sentinel[slice_with_sentinel.len] == 0);
}
```

### Expected Behavior

Test should generate a compile error, like it does on stage1:

```
/home/topolarity/repos/parselle/test4.zig:7:53: error: slice-sentinel does not match memory at target index
    const slice_with_sentinel: [:0]const u8 = string[0..8 :0];
                                                    ^
```

### Actual Behavior

```
1/1 test_0... thread 8107 panic: reached unreachable code
???:?:?: 0x20ba16 in ??? (???)
???:?:?: 0x20b9e4 in ??? (???)
???:?:?: 0x20cb4a in ??? (???)
???:?:?: 0x20c126 in ??? (???)
???:?:?: 0x20bc92 in ??? (???)
error: the following test command crashed:
../parselle/zig-cache/o/c17a70b385c9c4b8c89af09472844a2e/test /home/topolarity/repos/zig/stage2/bin/zig
```

**Comment by topolarity** [comment]
Also not sure why the panic handler fails to give a backtrace for this example

**Comment by martinhath** [comment]
It looks like this has been fixed.
```
$ build/zig test 12037.zig
./12037.zig:7:53: error: slice-sentinel does not match memory at target index
    const slice_with_sentinel: [:0]const u8 = string[0..8 :0];
                                                    ^
$ stage2/bin/zig test 12037.zig
12037.zig:7:53: error: value in memory does not match slice sentinel
    const slice_with_sentinel: [:0]const u8 = string[0..8 :0];
                                              ~~~~~~^~~~~~~~~
12037.zig:7:53: note: expected '0', found '111'
```

**Comment by Vexu** [comment]
Implemented in 02dc0732604236a57b43b9612d9b0571f06f905a
